src/TParsec/Combinators.idr

```diff
@@ -76,128 +146,342 @@ andbind p q = MkParser $ \mlen, ts =>
 randm : Monad mn =>
         All (Parser mn p a :-> Cst (mn b) :-> Parser mn p b)
 randm p q = map snd (andm p q)
 
+||| Run a monadic computation and a parser in succession.
+|||
+||| Given a monad `M`, a monadic computation `M a` and a parser `Parser M a`
+||| this function will run the comutation `M a` first and then run the 
+||| parser. Both results are returned as a pair.
+|||
+||| Unindexed signature: `M a -> Parser M b -> Parser M (a, b)`
 mand : Monad mn =>
        All (Cst (mn a) :-> Parser mn p b :-> Parser mn p (a, b))
 mand p q = MkParser $ \mlen, ts => do a <- p
                                       Functor.map (Success.map (MkPair a)) (runParser q mlen ts)
 
+||| Like `mand` but discards the second value.
+|||
+||| Given a monad `M`, a monadic computation `M a` and a parser `Parser M a`
+||| this function will run the comutation `M a` first and then run the 
+||| parser. The result of the parser is discarded.
+|||
+||| Unindexed signature: `M a -> Parser M b -> Parser M a`
 lmand : Monad mn =>
         All (Cst (mn a) :-> Parser mn p b :-> Parser mn p a)
 lmand p q = map fst (mand p q)
 
+||| Like `mand` but discards the first value.
+|||
+||| Given a monad `M`, a monadic computation `M a` and a parser `Parser M a`
+||| this function will run the comutation `M a` first and then run the 
+||| parser. The result of the computation is discarded.
+|||
+||| Unindexed signature: `M a -> Parser M b -> Parser M b`
 rmand : Monad mn =>
         All (Cst (mn a) :-> Parser mn p b :-> Parser mn p b)
 rmand p q = map snd (mand p q)
 
+||| Runs two parsers in succession but the second one is allowed to fail.
+|||
+||| Runs two parsers one after the other, this parser succeeds as long
+||| as the first parser succeeds. If the second one fails, its value
+||| will be reported as `Nothing`. Both values are returned as a pair.
+|||
+||| Unindexed signature: `Parser a -> Parser b -> Parser (a, Maybe b)`
 andopt : (Monad mn, Alternative mn) =>
          All (Parser mn p a :-> Box (Parser mn p b) :-> Parser mn p (a, Maybe b))
 andopt p q = andoptbind p (\ _ => q)
 
+||| Runs two parsers in succession but the first one is allowed to fail.
+|||
+||| Runs two parsers one after the other, this parser succeeds as long
+||| as the second parser succeeds. If the first one fails, its value
+||| will be reported as `Nothing`. Both values are returned as a pair.
+|||
+||| Unindexed signature: `Parser a -> Parser b -> Parser (Maybe a, b)`
 optand : (Monad mn, Alternative mn) =>
          All (Parser mn p a :-> Parser mn p b :-> Parser mn p (Maybe a, b))
 optand p q = alt (and (map Just p) q) (map (MkPair Nothing) q)
 
+||| Parse a value and processes it into another parser.
+|||
+||| Given a `Parser a` and a function `a -> Parser b` this function will 
+||| attempt the first parser on the input, and run the function on the
+||| parsed value. If both those steps are successful, both values are
+||| returned as a pair.
+|||
```

Comment:
Only the second value is returned here as opposed to `andbind`.

Reply:
Fixed, can you tell me if this rewording is acceptable?

Reply:
:+1: